api/external_auth.proto

```diff
@@ -0,0 +1,152 @@
+    // If the IP address is private, the `region_code` should reflect the
+    // physical location where this peer is running.
+    string region_code = 7;
+  }
+
+  // This message defines attributes for an HTTP request. If the actual
+  // request is not an HTTP request, the runtime system should try to map
+  // the actual request to an equivalent HTTP request.
+  message Request {
+    // The unique ID for a request, which can be propagated to downstream
+    // systems. The ID should have low probability of collision
+    // within a single day for a specific service.
+    // For http it will be X-Request-ID or equivalent.
+    // For tcp is should be a connection id.
+    string id = 1;
+
+    // The HTTP request method, such as `GET`, `POST`.
+    string method = 2;
+
+    // The HTTP request headers. If multiple headers share the same key, they
+    // must be merged according to the HTTP spec. All header keys must be
+    // lowercased, because HTTP header keys are case-insensitive.
+    map<string, string> headers = 3;
+
+    // The HTTP URL path.
+    string path = 4;
+
+    // The HTTP request `Host` or 'Authority` header value.
+    string host = 5;
+
+    // The HTTP URL scheme, such as `http` and `https`.
+    string scheme = 6;
+
+    // The HTTP URL query in the format of `name1=value`&name2=value2`, as it
+    // appears in the first line of the HTTP request. No decoding is performed.
+    string query = 7;
+
+    // The HTTP URL fragment. No URL decoding is performed.
+    string fragment = 8;
+
+    // The timestamp when the `destination` service receives the first byte of
+    // the request.
+    google.protobuf.Timestamp time = 9;
+
+    // The HTTP request size in bytes. If unknown, it must be -1.
+    int64 size = 10;
+
+    // The network protocol used with the request, such as "http/1.1",
+    // "spdy/3", "h2", "h2c", "webrtc", "tcp", "udp", "quic". See
+    // https://www.iana.org/assignments/tls-extensiontype-values/tls-extensiontype-values.xhtml#alpn-protocol-ids
+    // for details.
+    string protocol = 11;
+  }
+
+  // The source of a network activity, such as starting a TCP connection.
+  // In a multi hop network activity, the source represents the sender of the
+  // last hop.
+  Peer source = 1;
+
+  // The destination of a network activity, such as accepting a TCP connection.
+  // In a multi hop network activity, the destination represents the receiver of
+  // the last hop.
+  Peer destination = 2;
+
+  // Represents a network request, such as an HTTP request.
+  Request request = 4;
+
+  // This is analogous to request.headers, however these contents will not be sent to the upstream server.
+  // request_context provides an extension mechanism for sending additional information to the auth server
+  // without adding it to request.headers.
+  // request_context may map to the internal opaque context in the filter chain.
+  map<string, string> request_context = 10;
```

Comment:
Since the extensions can be related to source or destination or connection or other parts, we should not use "request_context" as the name. I would just call it "extensions". And document the namespace:
- For proposed Envoy attributes, it should start with a tentative namespace, such as "request.auth" and call it "request.auth.???".
- For company specific attributes, it should start with a company name, such as "ibm.xxx." or "google.xxx".

Reply:
In Envoy, we already have a scheme for filter and metadata naming, where we follow reverse DNS. For Envoy internal things, we do "envoy.foo". For everything else, it's "com.google.foo", "com.ibm.foo". Should we be consistent with this? See second last point at https://github.com/envoyproxy/data-plane-api/blob/master/API_OVERVIEW.md#principles.